oximeter/db/src/oxql/ast/table_ops/group_by.rs

```diff
@@ -0,0 +1,208 @@
+use crate::oxql::Table;
+use anyhow::Context;
+use num::ToPrimitive;
+use std::collections::BTreeMap;
+
+/// A table operation for grouping data by fields, apply a reducer to the
+/// remaining.
+#[derive(Clone, Debug, PartialEq)]
+pub struct GroupBy {
+    pub identifiers: Vec<Ident>,
+    pub reducer: Reducer,
+}
+
+impl GroupBy {
+    // Apply the group_by table operation.
+    pub(crate) fn apply(&self, tables: &[Table]) -> Result<Vec<Table>, Error> {
+        anyhow::ensure!(
+            tables.len() == 1,
+            "Group by operations require exactly one table",
+        );
+        let table = &tables[0];
+        let mut output_table = Table::new(table.name());
+        let kept_fields: Vec<_> =
+            self.identifiers.iter().map(Ident::as_str).collect();
+        let mut counts: BTreeMap<_, u64> = BTreeMap::new();
+        for input in table.iter() {
+            anyhow::ensure!(
+                input.points.len() > 0,
+                "Timeseries cannot be empty"
+            );
+
+            // For now, we can only apply this to 1-D timeseries.
+            anyhow::ensure!(
+                input.points.dimensionality() == 1,
+                "Group-by with multi-dimensional timeseries is not yet supported"
+            );
+            let data_type = input.points.data_types().next().unwrap();
+            anyhow::ensure!(
+                data_type.is_numeric(),
+                "Only numeric data types can be grouped, not {}",
+                data_type,
+            );
+            let metric_type = input.points.metric_types().next().unwrap();
+            anyhow::ensure!(
+                !matches!(metric_type, MetricType::Cumulative),
+                "Cumulative metric types cannot be grouped",
+            );
+            anyhow::ensure!(
+                table.is_aligned(),
+                "Input tables to a `group_by` must be aligned"
+            );
+
+            // Throw away the fields in this timeseries that are not in the
+            // group_by list.
+            let dropped = input.copy_with_fields(&kept_fields)?;
+            let key = dropped.key();
+
+            // Fetch the existing table, if one exists. If one does _not_ exist,
+            // we'll insert the table with the data type converted to a thee
+            // appropriate type. For a mean reducer, the output is always a
+            // double. For others, the value may be the same as the input type.
+            //
+            // TODO-completeness: We need to support reducers other than `Mean`
+            // and `Sum` here.
+            match output_table.get_mut(key) {
+                Some(existing) => {
+                    anyhow::ensure!(
+                        existing.points.len() == dropped.points.len(),
+                        "Cannot group timeseries with different numbers of data points",
+                    );
+                    assert!(
+                        existing.points.timestamps == dropped.points.timestamps
```

Comment:
So I've been looking at this for a while, and I think the whole thing is just wrong. During development, I made the short cut of simply merging arrays together by index during group-by operations. Then I implemented the right thing for `join`s (stacking elements by timestamp, not index), but I didn't go back and do that for `group_by` as well. I didn't notice because most of the metrics I was grouping have nearly identical timepoints to begin with (e.g., the physical data link metrics, which all derive from `kstat`s sampled at the same time).

I'll make this a proper time-based merge and add some more tests. Thanks for catching.

Reply:
Ok, 99b921f82 implements what I think is a correct approach to grouping. That matches output data points by timestamp, not index. It's possibly a bit funky since we also filter missing data, rather than grouping it. I want to think more about exactly how to handle that case, but I need to run for now. I might update this later to keep the missing value in the output. It's a bit tricky, since we only want to do that if it's the _only_ missing value. I'm not sure if that's the right semantics for the `group_by` operation, I need to think a bit about it.

Anyway, thanks for flagging this @rcgoodfellow, this is much better.

Reply:
Alright, ff13145fd is better. The difference from the latter commit is how we handle missing values in group-by and align operations. Before, they were filled in with a zero-value during alignment, and left off entirely in grouping. In this latest commit, we do not "overwrite" them in either case. If there is an alignment period with only missing values, the output also has only a missing value. If there is a period with only missing values in a group by, the output has only a missing value.

This makes things more consistent and easier to predict. Missing values don't get removed during grouping, or zero-filled in alignment -- they're slightly sticky, in that you need a _non-missing_ value to take precedence.